cpp/src/parquet/column_reader.h

```diff
@@ -302,8 +303,274 @@ class TypedColumnReader : public ColumnReader {
+    Range rangeToAdd = range;
+    if (merge) {
+      for (int i = static_cast<int>(ranges.size()) - 1; i >= 0; --i) {
+        Range last = ranges[i];
+        if (last.isAfter(range)) {
+          throw ParquetException(range.toString() + " cannot be added to " +
+                                 this->toString());
+        }
+        const Range u = Range::unionRange(last, rangeToAdd);
+        if (u.from == -1 && u.to == -1) {
+          break;
+        }
+        rangeToAdd = u;
+        ranges.erase(ranges.begin() + i);
+      }
+    } else {
+      if (ranges.size() > 1) assert(rangeToAdd.from > ranges.back().to);
+    }
+    ranges.push_back(rangeToAdd);
+  }
+
+  size_t rowCount() const {
+    size_t cnt = 0;
+    for (const Range& range : ranges) {
+      cnt += range.count();
+    }
+    return cnt;
+  }
+
+  bool isValid() const {
+    if (ranges.size() == 0) {
+      return false;
+    }
+    if (ranges[0].from < 0) {
+      return false;
+    }
+    for (size_t i = 1; i < ranges.size(); i++) {
+      if (ranges[i].from <= ranges[i - 1].to) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool isOverlapping(int64_t from, int64_t to) const {
+    const Range searchRange(from, to);
+    return isOverlapping(searchRange);
+  }
+
+  bool isOverlapping(const Range& searchRange) const {
+    auto it = std::lower_bound(
+        ranges.begin(), ranges.end(), searchRange,
+        [](const Range& r1, const Range& r2) { return r1.isBefore(r2); });
+    return it != ranges.end() && !(*it).isAfter(searchRange);
+  }
+
+  std::vector<Range>& getRanges() { return ranges; }
+
+  const Range& operator[](size_t index) const { return ranges[index]; }
+
+  std::string toString() const {
+    std::string result = "[";
+    for (const Range& range : ranges) {
+      result +=
+          "(" + std::to_string(range.from) + ", " + std::to_string(range.to) + "), ";
+    }
+    if (!ranges.empty()) {
+      result = result.substr(0, result.size() - 2);
+    }
+    result += "]";
+    return result;
+  }
+};
+
+using RowRangesPtr = std::shared_ptr<RowRanges>;
+
 namespace internal {
 
+class PARQUET_EXPORT RecordSkipper {
+ public:
+  RecordSkipper(RowRanges& pages, RowRanges& row_ranges_)
+      : row_ranges(row_ranges_) {  // copy row_ranges
+    RowRanges will_process_pages, skip_pages;
+    for (auto& page : pages.getRanges()) {
+      if (!row_ranges.isOverlapping(page)) {
+        skip_pages.add(page, false);
+      }
+    }
+
+    /// Since the skipped pages will be slienly skipped without updating
+    /// current_rg_processed_records or records_read_, we need to pre-process the row
+    /// ranges as if these skipped pages never existed
+    adjust_ranges(skip_pages, row_ranges);
+
+    total_rows_to_process = pages.rowCount() - skip_pages.rowCount();
+  }
+
+  /// \brief Return the number of records to read or to skip
+  /// if return values is positive, it means to read N records
```

Comment:
is sign reversal really necessary?  Couldn't the reader we have a conention were the first element returned is always  records to read and then alternate between inclusion and exclusion?